File: AIOUSB/samples/USB-DIO-32/Sample.java
/**
 * @file   Sample.java
 * @author $Format: %an <%ae>$
 * @date   $Format: %ad$
 * @release $Format: %t$
 * @brief  
 * @note to compile:
 *   make Sample.java
 */

import AIOUSB.*;
import java.io.*;
import java.util.*;
import java.util.concurrent.*;

class Sample {
    public static final String VERSION = "$Format: %t$";
    /*
     * if you want to test writing to the EEPROM, set DEMO_EEPROM_WRITE to 'true'; after writing
     * to the EEPROM this sample program will attempt to restore the original EEPROM contents,
     * but you never know ...
     */
    static class Device {
        public int deviceMask;
        public SWIGTYPE_p_unsigned_long productID  ;
        public SWIGTYPE_p_unsigned_long nameSize   ;
        public SWIGTYPE_p_unsigned_long numDIOBytes;
        public SWIGTYPE_p_unsigned_long numCounters;            
        public AIOChannelMask outputMask;
        int MAX_DIO_BYTES = 32;
        public DIOBuf readBuffer;
        public DIOBuf writeBuffer;
        public long index;
        public Device( long nindex, 
                       SWIGTYPE_p_unsigned_long nproductID,  
                       SWIGTYPE_p_unsigned_long nnameSize,  
                       SWIGTYPE_p_unsigned_long nnumDIOBytes, 
                       SWIGTYPE_p_unsigned_long nnumCounters
                       ) { 
            index = nindex;
            // System.out.println("Used device " + index );
            productID = nproductID;
            nameSize = nnameSize;
            numDIOBytes = nnumDIOBytes;
            numCounters = nnumCounters;
            outputMask   = new AIOChannelMask(0);
            readBuffer   = new DIOBuf(MAX_DIO_BYTES );
            writeBuffer  = new DIOBuf( MAX_DIO_BYTES );
        }
    }

    protected static final boolean DEMO_EEPROM_WRITE = false;
    
    public static void main( String args[] ) {

        System.out.println(
                           "USB-DIO-32 sample program version: " + VERSION + "\n" + 
                           "  AIOUSB Java library version: " + 
                           "  AIOUSB library version: " + 
                           AIOUSB.AIOUSB_GetVersion() + ", " + 
                           AIOUSB.AIOUSB_GetVersionDate() + "\n" + 
                           "  JRE version: " + System.getProperty( "java.version" ) + "\n" +
                           "  OS version: " + System.getProperty( "os.name" ) + 
                           " " + System.getProperty( "os.arch" ) + 
                           " " + System.getProperty( "os.version" ) + "\n" + 
                           "  This program demonstrates controlling a USB-DIO-32 device on\n" + 
                           "  the USB bus. For simplicity, it uses the first such device found\n" + 
                           "  on the bus."
                           );
        
        AIOUSB.AIOUSB_Init();
        AIOUSB.AIOUSB_ListDevices();
        int number_devices = 1;
        long deviceMask = AIOUSB.GetDevices();
        DIOBuf buf = new DIOBuf(32);

        // DIOBuf buf = AIOUSB.NewDIOBufFromBinStr("10101010001100111111000011111111" );
        // char val = 0xff;
        // AIOUSB.DIOBufSetByteAtIndex( buf,  (long)1, (char)0xff );
        // System.out.println(AIOUSB.DIOBufToString(buf));

        SWIGTYPE_p_unsigned_long productID    = AIOUSB.new_ulp();
        SWIGTYPE_p_unsigned_long nameSize     = AIOUSB.new_ulp();
        SWIGTYPE_p_unsigned_long numDIOBytes  = AIOUSB.new_ulp();
        SWIGTYPE_p_unsigned_long numCounters  = AIOUSB.new_ulp();
        String name = "";
        long index = 0;
        ArrayList<Device> devices = new ArrayList<Device>();

        while( deviceMask > 0 && (devices.size() < number_devices  ) ) {
            if ( (deviceMask & 1 ) != 0 ) { 
                AIOUSB.QueryDeviceInfo(index , productID, nameSize, name, numDIOBytes, numCounters);
                if( AIOUSB.ulp_value(productID) == ProductIDS.USB_DIO_32.swigValue() ) { 
                    System.out.println("Found device at index:" + index );
                    devices.add( new Device( index, productID, nameSize, numDIOBytes, numCounters ) );
                }
                index += 1;
                deviceMask >>= 1;
            }        
        }
        if( devices.size() != number_devices ) {
            System.out.println("Error; not enough devices found=" + number_devices + "\n" );
            System.out.println("Error; not enough devices found=" + devices.size() + "\n" );
            System.exit(1);
        } else {
            System.out.println(devices.size() + " devices were found\n");
        }

        Device device = devices.get(0);
        // System.out.println("Using device index:" + device.index  );

        AIOUSB.AIOUSB_SetCommTimeout( device.index, 1000 );
        device.outputMask = new AIOChannelMask(0);

        // Use the following setting to allow writing to each of the 4 banks of 8 channels
        AIOUSB.AIOChannelMaskSetMaskFromStr( device.outputMask, "1111" );

        for( int port = 0; port < 0x20; port ++ ) { 
            System.out.println( "Using value " + port );
            AIOUSB.DIOBufSetIndex( device.writeBuffer, port, 1 );
            long result = AIOUSB.DIO_ConfigureWithDIOBuf( device.index, (short)0 , device.outputMask , device.writeBuffer );
            if (result != 0 ) { 
                System.out.println("\tresult value:" + result );
            }
            try {
                TimeUnit.MILLISECONDS.sleep(100);// Thread.sleep(0,10000);
                AIOUSB.DIO_ReadAllToDIOBuf(0, buf);
                System.out.println( buf.toString() );
            } catch ( java.lang.InterruptedException e ) {

            }
        }
        long result;
        try { 
            // ONLY Port D
            AIOUSB.AIOChannelMaskSetMaskFromStr(device.outputMask, "1000" );
            result = AIOUSB.DIO_ConfigureWithDIOBuf( device.index, (short)0, device.outputMask , device.writeBuffer );
            TimeUnit.MILLISECONDS.sleep(500);
            // ONLY Port C        
            AIOUSB.AIOChannelMaskSetMaskFromStr(device.outputMask, "0100" );
            result = AIOUSB.DIO_ConfigureWithDIOBuf( device.index, (short)0, device.outputMask , device.writeBuffer );
            TimeUnit.MILLISECONDS.sleep(500);
            // ONLY Port B        
            AIOUSB.AIOChannelMaskSetMaskFromStr(device.outputMask, "0010" );
            result = AIOUSB.DIO_ConfigureWithDIOBuf( device.index, (short)0, device.outputMask , device.writeBuffer );
            TimeUnit.MILLISECONDS.sleep(500);
            // ONLY Port A   
            AIOUSB.AIOChannelMaskSetMaskFromStr(device.outputMask, "0001" );
            result = AIOUSB.DIO_ConfigureWithDIOBuf( device.index, (short)0, device.outputMask , device.writeBuffer );
            TimeUnit.MILLISECONDS.sleep(500);
        } catch ( java.lang.InterruptedException e ) {

        }

    }
}

File: AIOUSB/samples/USB-DIO-32/sample.py
#!/usr/bin/env python
# @file   sample.py
# @author $Format: %an <%ae>$
# @date   $Format: %ad$
# @release $Format: %t$
# @ingroup samples
# @brief Python version of the C/C++ samples that demonstrate reading / writing from the DIO-32 card
#
#/

import sys
import time
from AIOUSB import *

MAX_DIO_BYTES = 32
#
# Simple class for keeping track of deviecs found
#
class Device:
    outputMask = NewAIOChannelMaskFromStr("1111")
    readBuffer = DIOBuf(MAX_DIO_BYTES )
    writeBuffer = DIOBuf( MAX_DIO_BYTES )
    name = ""
    serialNumber = 0
    index = 0
    numDIOBytes = 0
    numCounters = 0
    productID = 0
    def __init__(self, **kwds):
        self.__dict__.update(kwds)


devices = []                    # Array of our Devices
number_devices = 1

devices = []                    # Array of our Devices
number_devices = 1

print """
USB-DIO-96 sample program version %s, %s 
This program demonstrates communicating using the USB-DIO-96 product.
AIOUSB library version %s, %s the same USB bus. It uses the 
first device found found on the bus
""" % ( "$Format: %t$", "$Format: %ad$", AIOUSB_GetVersion(), AIOUSB_GetVersionDate() )


result = AIOUSB_Init()
if result != AIOUSB_SUCCESS:
    sys.exit("Unable to initialize USB devices")
devicesFound = 0
deviceMask = GetDevices()
index = 0
AIOUSB_ListDevices()       



while deviceMask > 0 and len(devices) < number_devices :
    if (deviceMask & 1 ) != 0:
        obj = AIODeviceInfoGet( index )
        if obj.PID == USB_DIO_32 or obj.PID == USB_DIO_16A :
            devices.append( Device( index=index, productID=obj.PID, numDIOBytes=obj.DIOBytes,numCounters=obj.Counters ))
    index += 1
    deviceMask >>= 1
try:
    device = devices[0]
except IndexError:
    print """No devices were found. Please make sure you have at least one 
ACCES I/O Products USB device plugged into your computer"""
    sys.exit(1)



AIOUSB_SetCommTimeout( device.index, 1000 )
AIOChannelMaskSetMaskFromStr( device.outputMask, "1111" )
print "Mask was %s " % ( AIOChannelMaskToString( device.outputMask ))

buf = DIOBuf(32)

for port in range(0x20):
    print "Using value %d" % (port)
    DIOBufSetIndex( device.writeBuffer, port, 1 );
    result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask, device.writeBuffer )
    DIO_ReadAllToDIOBuf(0, buf)
    print buf
    #print "Result was %d" % (result)
    time.sleep(1/10.0)

# ONLY Port D
AIOChannelMaskSetMaskFromStr(device.outputMask, "1000" );
result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask , device.writeBuffer );
time.sleep(1/10.0)
# ONLY Port C
AIOChannelMaskSetMaskFromStr(device.outputMask, "0100" );
result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask , device.writeBuffer );
time.sleep(1/10.0);
# ONLY Port B
AIOChannelMaskSetMaskFromStr(device.outputMask, "0010" );
result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask , device.writeBuffer );
time.sleep(1/10.0);
# ONLY Port A
AIOChannelMaskSetMaskFromStr(device.outputMask, "0001" );
result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask , device.writeBuffer );
time.sleep(1/10.0);
# All Channels
AIOChannelMaskSetMaskFromStr(device.outputMask, "1111" );
result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask , device.writeBuffer );

tmpbuf = list("00000000000000000000000000000000")

for i in range(len(tmpbuf)):
    if i >= 1:
        tmpbuf[i] = "0"

    tmpbuf[i] = '1'
    DIOBufReplaceBinString(device.writeBuffer, "".join(tmpbuf) )
    result = DIO_ConfigureWithDIOBuf( device.index, AIOUSB_FALSE, device.outputMask , device.writeBuffer )    
    time.sleep(1/10.0)


